### PR TITLE
Run system-probe tests using kernel matrix testing scenario

### PR DESCRIPTION
**Pull request by usamasaqib**
### What does this PR do?
This PR finalizes the integration of the new kernel-matrix-testing scenario in the datadog-agent CI.
The PR adds support to launch the environment for running the tests, and adds the skeleton for the jobs to launch the tests themselves. The instances are launched in the datadog-agent-qa tenant.
Subsequent commits will add support for various ancillary features such as integration with the CI visibility product.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 1]
```diff
@@ -159,4 +190,25 @@ kernel_matrix_testing_upload_deps:
...
+  variables:
+    AWS_EC2_SSH_KEY_FILE: $CI_PROJECT_DIR/ssh_key
+    GOVERSION: "1.19"
+  before_script:
+    - !reference [.kernel_matrix_testing_new_profile]
+  script:
+    - INSTANCE_IP=$(cat $CI_PROJECT_DIR/stack.outputs | grep $ARCH-instance-ip | cut -d ' ' -f 2)
+    - MICRO_VM_IP=$(cat $CI_PROJECT_DIR/stack.outputs | grep $ARCH-$TAG | cut -d ' ' -f 2)
+    - chmod 600 $AWS_EC2_SSH_KEY_FILE
+    # Without the newline ssh silently fails and moves on to try other auth methods
```
Comment (pducolin) [inline .gitlab/kernel_version_testing/system_probe.yml:207]: Interesting

File: test/new-e2e/system-probe/system-probe-test-env.go [hunk 2]
```diff
@@ -44,42 +50,55 @@ type TestEnv struct {
...
+		context: context.Background(),
+		name:    fmt.Sprintf("microvm-scenario-%s", name),
 	}
-	defer f.Close()
-	f.WriteString(sshkey)
 
 	stackManager := infra.GetStackManager()
 
 	config := runner.ConfigMap{
+		"ddinfra:env":                            auto.ConfigValue{Value: opts.InfraEnv},
```
Comment (pducolin) [inline test/new-e2e/system-probe/system-probe-test-env.go:98]: **💬 suggestion**
We should discuss about possibly moving this to a `TestSuite` that `go test` could run, vs using `go run` with a custom executable go program.
We are moving towards `TestSuite` to help handling configmap / upresult parsing. 

In the meantime, I suggest using keys from test_infra_definitions - you can find an example is https://github.com/DataDog/datadog-agent/blob/main/test/new-e2e/runner/configmap.go#L20-L24

Could you consider using the configmap from test helpers, in a follow up ?

File: test/new-e2e/system-probe/system-probe-test-env.go [hunk 3]
```diff
@@ -90,84 +109,103 @@ func NewTestEnv(name, securityGroups, subnets, x86InstanceType, armInstanceType
 		"microvm:arm64AmiID":                     auto.ConfigValue{Value: opts.ArmAmiID},
 		"microvm:workingDir":                     auto.ConfigValue{Value: CustomAMIWorkingDir},
 	}
+	// We cannot add defaultPrivateKeyPath if the key is in ssh-agent, otherwise passphrase is needed
```
Comment (pducolin) [inline test/new-e2e/system-probe/system-probe-test-env.go:112]: **💬 suggestion**
You can skip passphrase on local environment, on macOs, registering the key with apple key chain

```bash
ssh-add --apple-use-keychain <path/to/private_key>
```

File: test/new-e2e/system-probe/system-probe-test-env.go [hunk 4]
```diff
@@ -90,84 +109,103 @@ func NewTestEnv(name, securityGroups, subnets, x86InstanceType, armInstanceType
...
 		return nil
-	}, opts.FailOnMissing)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create stack: %w", err)
+	}); retryErr != nil {
+		return nil, fmt.Errorf("failed to create stack: %w", retryErr)
 	}
 
-	b, err := json.MarshalIndent(upResult.Summary, "", "	")
+	err = outputsToFile(upResult.Outputs)
```
Comment (pducolin) [inline test/new-e2e/system-probe/system-probe-test-env.go:202]: **💬 suggestion**
This should be handled by a deserialiser - as soon as we document it, good for now

**Review by pducolin — APPROVED** [review]
Approving this, but as soon as we document `TestSuite` it should be adapted to have uniform test design in `new-e2e` tests.

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 5]
```diff
@@ -99,29 +103,46 @@ package_dependencies_arm64:
...
+    INFRA_ENV: "aws/agent-qa"
+  before_script:
+    - !reference [.retrieve_linux_go_deps]
+    - !reference [.kernel_matrix_testing_new_profile]
   script:
-    - echo s3://dd-ci-datadog-agent-omnibus-cache-build-stable/kernel-matrix-testing/$CI_PIPELINE_ID/$CI_JOB_ID/pulumi > $STACK_DIR
+    - echo "s3://dd-pulumi-state?region=us-east-1&awssdk=v2&profile=$AWS_PROFILE" > $STACK_DIR
     - pulumi login $(cat $STACK_DIR | tr -d '\n')
-    - inv -e system-probe.test-microvms --security-groups=$KITCHEN_EC2_SG_IDS --subnets=$KITCHEN_EC2_SUBNET --instance-type-x86=$EC2_X86_INSTANCE_TYPE --instance-type-arm=$EC2_ARM_INSTANCE_TYPE --x86-ami-id=$X86_AMI_ID --arm-ami-id=$ARM_AMI_ID
+    - aws ssm get-parameter --region us-east-1 --name ci.datadog-agent.ssh_key --with-decryption --query "Parameter.Value" --out text > $AWS_EC2_SSH_KEY_FILE
```
Comment (brycekahle) [inline .gitlab/kernel_version_testing/system_probe.yml:143]: Does this need to be wrapped with `set +x/set -x`?

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 6]
```diff
@@ -131,24 +152,34 @@ kernel_matrix_testing_setup_env:
...
+      - $AWS_EC2_SSH_KEY_FILE
 
 kernel_matrix_testing_upload_deps:
   extends:
     - .kitchen_ec2_location_us_east_1
   stage: kernel_matrix_testing
-  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-buildimages/system-probe_x64:$DATADOG_AGENT_SYSPROBE_BUILDIMAGES
-  needs: ["kernel_matrix_testing_setup_env", "package_dependencies_x64", "package_dependencies_arm64"]
-  tags: ["runner:main"]
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/test-infra-definitions/runner:7c39524b
```
Comment (brycekahle) [inline .gitlab/kernel_version_testing/system_probe.yml:162]: Should the image hash be a variable?

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 7]
```diff
@@ -159,4 +190,25 @@ kernel_matrix_testing_upload_deps:
...
+    AWS_EC2_SSH_KEY_FILE: $CI_PROJECT_DIR/ssh_key
+    GOVERSION: "1.19"
+  before_script:
+    - !reference [.kernel_matrix_testing_new_profile]
+  script:
+    - INSTANCE_IP=$(cat $CI_PROJECT_DIR/stack.outputs | grep $ARCH-instance-ip | cut -d ' ' -f 2)
+    - MICRO_VM_IP=$(cat $CI_PROJECT_DIR/stack.outputs | grep $ARCH-$TAG | cut -d ' ' -f 2)
+    - chmod 600 $AWS_EC2_SSH_KEY_FILE
+    # Without the newline ssh silently fails and moves on to try other auth methods
+    - echo "" >> $AWS_EC2_SSH_KEY_FILE
```
Comment (brycekahle) [inline .gitlab/kernel_version_testing/system_probe.yml:208]: Should we do this when creating the file? That way it is ready to be used everywhere.

File: tasks/system_probe.py [hunk 8]
```diff
@@ -1527,24 +1527,26 @@ def save_test_dockers(ctx, output_dir, arch, windows=is_windows):
...
     args = [
-        f"--sgs {security_groups}" if security_groups is not None else "",
-        f"--subnets {subnets}" if subnets is not None else "",
         f"--instance-type-x86 {instance_type_x86}" if instance_type_x86 is not None else "",
         f"--instance-type-arm {instance_type_arm}" if instance_type_arm is not None else "",
         f"--x86-ami-id {x86_ami_id}" if x86_ami_id is not None else "",
         f"--arm-ami-id {arm_ami_id}" if arm_ami_id is not None else "",
         "--destroy" if destroy else "",
         "--upload-dependencies" if upload_dependencies else "",
+        f"--ssh-key-path {ssh_key_path}" if ssh_key_path is not None else "",
```
Comment (brycekahle) [inline tasks/system_probe.py:1547]: I *think* this could be simplified to this:
```suggestion
        f"--ssh-key-path {ssh_key_path}" if ssh_key_path else "",
```

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 9]
```diff
@@ -159,4 +190,25 @@ kernel_matrix_testing_upload_deps:
...
+    # ssh into each micro-vm and run initialization script. This script will also run the tests.
+    - ssh -o StrictHostKeyChecking=no -i $AWS_EC2_SSH_KEY_FILE ubuntu@$INSTANCE_IP "ssh -o StrictHostKeyChecking=no -i /home/kernel-version-testing/ddvm_rsa root@${MICRO_VM_IP} 'cp /opt/kernel-version-testing/dependencies-${ARCH}.tar.gz / && cd / && tar xzf dependencies-${ARCH}.tar.gz --strip-components=1 && ls -la / && chmod +x /micro-vm-init.sh && /micro-vm-init.sh ${GOVERSION}'"
```
Comment (brycekahle) [inline .gitlab/kernel_version_testing/system_probe.yml:210]: If the init commands get any longer, they will probably need to be in a script file.

File: test/new-e2e/system-probe/test/micro-vm-init.sh [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+eval $(gimme $GOVERSION)
+## Start docker
+systemctl start docker
+## Load docker images
+find $KITCHEN_DOCKERS -maxdepth 1 -type f -exec docker load -i {} \;
+
+# VM provisioning end !
+
+# Start tests
+/system-probe-test_spec
```
Comment (brycekahle) [inline test/new-e2e/system-probe/test/micro-vm-init.sh:19]: I *think* you want `exec` here

```suggestion
exec /system-probe-test_spec
```

File: test/new-e2e/system-probe/test/system-probe-test_spec.go [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+		return nil
+	})
+	if err != nil {
+		return []string{}, err
+	}
+
+	return matches, nil
+}
+
+func generatePacakgeName(file string) string {
```
Comment (brycekahle) [inline test/new-e2e/system-probe/test/system-probe-test_spec.go:111]:
```suggestion
func generatePackageName(file string) string {
```

File: test/new-e2e/system-probe/test/system-probe-test_spec.go [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+	inclusive: false,
+}
+
+func pathEmbedded(fullPath, embedded string) bool {
+	normalized := fmt.Sprintf("/%s/",
+		strings.TrimRight(
+			strings.TrimLeft(embedded, "/"),
+			"/",
+		),
+	)
```
Comment (brycekahle) [inline test/new-e2e/system-probe/test/system-probe-test_spec.go:75]:
```suggestion
	normalized := fmt.Sprintf("/%s/", strings.Trim(embedded, "/"))
```

File: test/new-e2e/system-probe/test/system-probe-test_spec.go [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+	}
+
+	return mergedEnv
+}
+
+func runCommandAndStreamOutput(cmd *exec.Cmd, commandOutput io.Reader) error {
+	go func() {
+		scanner := bufio.NewScanner(commandOutput)
+		for scanner.Scan() {
+			_, _ = os.Stdout.Write([]byte(scanner.Text() + "\n"))
```
Comment (brycekahle) [inline test/new-e2e/system-probe/test/system-probe-test_spec.go:162]: It seems like you are just passing the text straight to `stdout`, so why is this necessary?

Comment (usamasaqib) [reply test/new-e2e/system-probe/test/system-probe-test_spec.go:162]: I have done this so we can stream the output rather than waiting for the command to complete before the output is printed.

File: .gitlab/kernel_version_testing/system_probe.yml (hunk 5 above)
Comment (usamasaqib) [reply .gitlab/kernel_version_testing/system_probe.yml:143]: I do not think so. Firstly the output is being redirected and it does not get printed.
Secondly this same pattern is used elsewhere as well for example [this](https://github.com/DataDog/datadog-agent/blob/main/.gitlab/functional_test/system_probe.yml#L22) job call run-test-kitchen which does the [same](https://github.com/DataDog/datadog-agent/blob/main/test/kitchen/tasks/run-test-kitchen.sh#L98) thing.

File: .gitlab/kernel_version_testing/system_probe.yml (hunk 6 above)
Comment (usamasaqib) [reply .gitlab/kernel_version_testing/system_probe.yml:162]: The E2E squad is using a different version currently. I am can ask them if they would rather move to a global variable.

File: test/new-e2e/system-probe/system-probe-test-env.go (hunk 2 above)
Comment (usamasaqib) [reply test/new-e2e/system-probe/system-probe-test-env.go]: Can you elaborate which configmap you are referring to?

File: .gitlab/kernel_version_testing/system_probe.yml (hunk 5 above)
Comment (usamasaqib) [reply .gitlab/kernel_version_testing/system_probe.yml:143]: Alternatively this job does what you suggest: https://github.com/DataDog/datadog-agent/blob/main/.gitlab/e2e.yml#L131

I will move to this approach too.

File: test/new-e2e/system-probe/test/system-probe-test_spec.go (hunk 13 above)
Comment (brycekahle) [reply test/new-e2e/system-probe/test/system-probe-test_spec.go:154]: There is an easier way, you can do `cmd.Stdout = os.Stdout` and `cmd.Stderr = os.Stderr`

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 14]
```diff
@@ -159,4 +195,22 @@ kernel_matrix_testing_upload_deps:
...
+  variables:
+    AWS_EC2_SSH_KEY_FILE: $CI_PROJECT_DIR/ssh_key
+    GOVERSION: "1.19"
+  before_script:
+    - !reference [.kernel_matrix_testing_new_profile]
+  script:
+    - INSTANCE_IP=$(cat $CI_PROJECT_DIR/stack.outputs | grep $ARCH-instance-ip | cut -d ' ' -f 2)
+    - MICRO_VM_IP=$(cat $CI_PROJECT_DIR/stack.outputs | grep $ARCH-$TAG | cut -d ' ' -f 2)
+    # ssh into each micro-vm and run initialization script. This script will also run the tests.
+    - ssh -o StrictHostKeyChecking=no -i $AWS_EC2_SSH_KEY_FILE ubuntu@$INSTANCE_IP "ssh -o StrictHostKeyChecking=no -i /home/kernel-version-testing/ddvm_rsa root@${MICRO_VM_IP} 'bash /root/fetch_dependencies.sh && /micro-vm-init.sh ${GOVERSION}'"
```
Comment (brycekahle) [inline .gitlab/kernel_version_testing/system_probe.yml:212]: Where is `fetch_dependencies.sh` living?

Comment (usamasaqib) [reply .gitlab/kernel_version_testing/system_probe.yml:212]: It is added in the micro-vm image:
https://github.com/DataDog/ami-builder/blob/usama.saqib/kernel-version-testing-ami/scripts/kernel-version-testing/make-dev-image.sh#L224
https://github.com/DataDog/ami-builder/blob/usama.saqib/kernel-version-testing-ami/scripts/kernel-version-testing/ubuntu/provision-cloudimg.sh#L184

File: tasks/system_probe.py [hunk 15]
```diff
@@ -1527,27 +1527,33 @@ def save_test_dockers(ctx, output_dir, arch, windows=is_windows):
...
     instance_type_x86=None,
     instance_type_arm=None,
     x86_ami_id=None,
     arm_ami_id=None,
     destroy=False,
     upload_dependencies=False,
+    ssh_key_name=None,
+    ssh_key_path=None,
+    dependencies_dir=None,
+    shutdown_period=320,
```
Comment (brycekahle) [inline tasks/system_probe.py:1540]: nit: this isn't a period, since it isn't periodic. It is more of a delay. I also think we shouldn't leave this as seemingly unitless (I realize it is minutes, but that isn't communicated here). I tried to change it to `time.Duration`, but that also involved changing the test definition. Let's fix this in a future PR.

File: test/new-e2e/system-probe/test/micro-vm-init.sh [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+eval $(gimme $GOVERSION)
+## Start docker
+systemctl start docker
+## Load docker images
+find $KITCHEN_DOCKERS -maxdepth 1 -type f -exec docker load -i {} \;
+
+# VM provisioning end !
+
+# Start tests
+IP=$(ip -f inet addr show $(ip route get $(getent ahosts google.com | awk '{print $1; exit}') | grep -Po '(?<=(dev ))(\S+)') | sed -En -e 's/.*inet ([0-9.]+).*/\1/p')
```
Comment (brycekahle) [inline test/new-e2e/system-probe/test/micro-vm-init.sh:19]: this long command should be documented, and possibly broken up into steps so it is clear what it is doing.

Comment (brycekahle) [reply test/new-e2e/system-probe/test/micro-vm-init.sh:19]: `$(getent ahosts google.com | awk '{print $1; exit}')` couldn't this be replaced with a known public IP like `8.8.8.8`?

Comment (brycekahle) [reply test/new-e2e/system-probe/test/micro-vm-init.sh]: I found a simpler version of this command

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 17]
```diff
@@ -159,4 +195,27 @@ kernel_matrix_testing_upload_deps:
...
+    - ssh -o StrictHostKeyChecking=no -i $AWS_EC2_SSH_KEY_FILE ubuntu@$INSTANCE_IP "ssh -o StrictHostKeyChecking=no -i /home/kernel-version-testing/ddvm_rsa root@${MICRO_VM_IP} 'bash /root/fetch_dependencies.sh ${ARCH} && /micro-vm-init.sh ${GOVERSION}'"
+    - ssh -o StrictHostKeyChecking=no -i $AWS_EC2_SSH_KEY_FILE ubuntu@$INSTANCE_IP "scp -o StrictHostKeyChecking=no -i /home/kernel-version-testing/ddvm_rsa root@{MICRO_VM_IP}:/junit-${MICRO_VM_IP}.tar.gz /home/ubuntu/junit-${MICRO_VM_IP}.tar.gz"
+    - ssh -o StrictHostKeyChecking=no -i $AWS_EC2_SSH_KEY_FILE ubuntu@$INSTANCE_IP "scp -o StrictHostKeyChecking=no -i /home/kernel-version-testing/ddvm_rsa root@{MICRO_VM_IP}:/testjson-${MICRO_VM_IP}.tar.gz /home/ubuntu/testjson-${MICRO_VM_IP}.tar.gz"
+    - scp -o StrictHostKeyChecking=no -i $AWS_EC2_SSH_KEY_FILE ubuntu@$INSTANCE_IP:/home/ubuntu/testjson-$MICRO_VM_IP.tar.gz $DD_AGENT_TESTING_DIR/kernel-matrix-testing-testjson-$MICRO_VM_NAME.tar.gz
+    - scp -o StrictHostKeyChecking=no -i $AWS_EC2_SSH_KEY_FILE ubuntu@$INSTANCE_IP:/home/ubuntu/junit-$MICRO_VM_IP.tar.gz $DD_AGENT_TESTING_DIR/kernel-matrix-testing-junit-$MICRO_VM_NAME.tar.gz
```
Comment (brycekahle) [inline .gitlab/kernel_version_testing/system_probe.yml:217]: Not required for this PR, but I'm wondering if we can create a simple ssh config file so we don't have to repeat all the options and proxy instructions each time.

File: tasks/system_probe.py [hunk 18]
```diff
@@ -1527,27 +1527,33 @@ def save_test_dockers(ctx, output_dir, arch, windows=is_windows):
...
+        f"--ssh-key-name {ssh_key_name}" if ssh_key_name else "",
+        f"--infra-env {infra_env}",
+        f"--shutdown-period {shutdown_period}",
+        f"--dependencies-dir {dependencies_dir}" if dependencies_dir else "",
     ]
 
     go_args = ' '.join(filter(lambda x: x != "", args))
     ctx.run(
-        f"cd ./test/new-e2e && go run ./scenarios/system-probe/main.go --name usama-saqib-test {go_args} --shutdown-period 720",
+        f"cd ./test/new-e2e && go run ./scenarios/system-probe/main.go --name usama-saqib-test {go_args}",
```
Comment (brycekahle) [inline tasks/system_probe.py:1558]: What is the significance of the `name` here? should it be hardcoded to `usama-saqib-test`?

File: test/new-e2e/system-probe/system-probe-test-env.go [hunk 19]
```diff
@@ -44,42 +55,55 @@ type TestEnv struct {
...
+	f, err := os.Create(stackOutputs)
+	if err != nil {
+		return fmt.Errorf("failed to create file: %s: %w", stackOutputs, err)
 	}
+	defer f.Close()
 
-	sshkey, err := runner.GetProfile().SecretStore().Get("aws_ec2_kitchen_ssh_key")
-	if err != nil {
-		return nil, fmt.Errorf("aws get credential: %w", err)
+	w := bufio.NewWriter(f)
```
Comment (brycekahle) [inline test/new-e2e/system-probe/system-probe-test-env.go:76]: I'm curious, why use a `bufio.Writer` here?

File: .gitlab/kernel_version_testing/system_probe.yml [hunk 20]
```diff
@@ -159,4 +196,28 @@ kernel_matrix_testing_upload_deps:
...
-
+kernel_matrix_testing_run_tests:
+  stage: kernel_matrix_testing
+  allow_failure: true
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/test-infra-definitions/runner:7c39524b
+  needs: ["kernel_matrix_testing_upload_deps", "kernel_matrix_testing_setup_env"]
+  tags: ["arch:amd64"]
+  variables:
+    AWS_EC2_SSH_KEY_FILE: $CI_PROJECT_DIR/ssh_key
+    GOVERSION: "1.19"
```
Comment (brycekahle) [inline .gitlab/kernel_version_testing/system_probe.yml:207]: This should probably be synced with the value in `.go-version`?

**Review by brycekahle — APPROVED** [review]
